server.go

```diff
@@ -43,14 +57,91 @@ func NewServer(connBufferSize int, httpPort, grpcPort int) *Server {
 //Init starts the server
 func (s *Server) Init() {
 	go s.listen()
-	s.fetchBigBang()
+	if s.StateFileName != "" {
+		err := s.loadState(filepath.Join(statesFolderName, s.StateFileName))
+		if err != nil {
+			fmt.Println("\nLoading state failed, exiting now", err)
+			panic(err)
+		}
+	} else {
+		s.fetchBigBang()
+	}
 }
 
 //Run offloads the computation of changes to cis
 func (s *Server) Run() {
+	signals := make(chan os.Signal, 1)
+	signal.Notify(signals, syscall.SIGINT, syscall.SIGTERM)
+
 	for {
+		if len(signals) != 0 {
+			s.grpcServer.Stop()
+			fmt.Println("Received Signal:", <-signals)
+			break
+		}
 		s.step()
 	}
+
+	err := s.saveState(time.Now())
+	if err == nil {
+		fmt.Println("\nState successfully saved")
+	} else {
+		fmt.Println("\nState could not be saved:", err)
+	}
```

Comment:
I would prefer the `s.saveState` and the `s.grpcServer.Stop` calls to be in a dedicated `*Server.Shutdown()` func

Reply:
why exporting the Shutdown func? we don't use it outside of server afaik

Reply:
yeah, it doesn't have to be exported